rpc/service.c

```diff
@@ -192,15 +192,63 @@ m0_rpc_service_reverse_session_get(struct m0_reqh_service   *service,
 	return M0_RC(rc);
 }
 
+static void rev_conn_free(struct m0_sm_group *grp, struct m0_sm_ast *ast)
+{
+	struct m0_reverse_connection *revc = ast->sa_datum;
+
+	M0_ENTRY("revc=%p", revc);
+	m0_rpc_link_fini(&revc->rcf_rlink);
+	rev_conn_tlink_del_fini(revc);
+	m0_free(revc);
+	M0_LEAVE("revc=%p", revc);
+}
+
+static bool rev_conn_disconnected_cb(struct m0_clink *link)
+{
+	struct m0_reverse_connection *revc;
+
+	revc = container_of(link, struct m0_reverse_connection, rcf_disc_wait);
+	revc->rcf_free_ast = (struct m0_sm_ast){
+		.sa_cb    = &rev_conn_free,
+		.sa_datum = revc,
+	};
+	m0_sm_ast_post(m0_locality0_get()->lo_grp, &revc->rcf_free_ast);
+	return true;
+}
+
+M0_INTERNAL void
+m0_rpc_service_reverse_session_put(struct m0_rpc_session *sess)
+{
+	struct m0_rpc_link           *rlk;
+	struct m0_reverse_connection *revc;
+
+	M0_ENTRY();
+
+	rlk  = container_of(sess, struct m0_rpc_link, rlk_sess);
+	revc = container_of(rlk, struct m0_reverse_connection, rcf_rlink);
+
+	M0_SET0(&revc->rcf_disc_wait);
+	if (revc->rcf_rlink.rlk_connected) {
+		m0_clink_init(&revc->rcf_disc_wait, rev_conn_disconnected_cb);
+		revc->rcf_disc_wait.cl_is_oneshot = true;
+		m0_rpc_link_disconnect_async(&revc->rcf_rlink,
+				m0_time_from_now(M0_REV_CONN_TIMEOUT, 0),
+				&revc->rcf_disc_wait);
+	}
```

Comment:
Is there a possibility that this function and the next cleanup() have some race conditions?
So they will be trying to call m0_rpc_link_disconnect_async() at the same time?

Reply:
Yes. Calling disconnect_async() at the same time - is probably not a problem. (Is it?) But cleaning up at the same time is problematic. That's why `m0_rpc_service_reverse_sessions_cleanup()` is called under the locality0 lock and `rev_conn_free()` is called as an AST under locality0.